`Development`: Add aspect to check user role authorization in course or for exercise (#8063)
ls1intum/Artemis: docs/dev/guidelines/server.rst

22. REST endpoint best practices for authorization
==================================================

To reject unauthorized requests as early as possible, Artemis employs two solutions:

> #. Implicit pre- and post-authorization annotations operate in two steps:
>     #. ``EnforceRoleInResource`` (e.g. ``EnforceRoleInCourse``) annotations perform pre authorization checks to block users with wrong or missing authorization roles without querying the database.
>     #. If necessary we check for access rights to individual resources within the database via light-weight queries.
>     #. Currently we offer the following annotations: ``EnforceRoleInCourse`` (e.g. ``EnforceAtLeastInstructorInCourse``), and ``EnforceRoleInExercise`` (e.g. ``EnforceAtLeastInstructorInExercise``)
> #. Explicit authorization checks:
>     #. ``PreAuthorize`` and ``Enforce`` annotations are responsible for blocking users with wrong or missing authorization roles without querying the database.
>     #. The ``AuthorizationCheckService`` is responsible for checking access rights to individual resources by querying the database.

Reviewer: Some change suggestions to make the code more readable (PreAuthorize was removed since it is only used for the EnforceAtLeastRole annotations)

```suggestion
#. Implicit pre- and post-authorization annotations:
    #. ``EnforceRoleInResource`` (e.g. ``EnforceAtLeastInstructorInCourse``) annotations are responsible for blocking users with wrong or missing authorization roles without querying the database.
    #. If necessary, these annotations check for access rights to individual resources within the database via light-weight queries.
    #. Currently we offer the following annotations: ``EnforceRoleInCourse`` and ``EnforceRoleInExercise``
#. Explicit authorization checks (which operate in two steps):
    #. ``EnforceAtLeastRole`` (e.g. ``EnforceAtLeastInstructor``) annotations are responsible for blocking users with wrong or missing authorization roles without querying the database.
    #. The ``AuthorizationCheckService`` is responsible for checking access rights to individual resources by querying the database.
```

> Because the first solution (Implicit pre- and post-authorizations) increase maintainability and is event faster in most cases, always annotate your REST endpoints with the corresponding ``EnforceRoleInResource`` annotation. Always use the annotation for the minimum role that has access.

Reviewer: Typos. Also, maybe it would make sense to re-structure this paragraph to put more focus on what to use:
**Always annotate your REST endpoints with the corresponding ``EnforceRoleInResource`` annotation (Implicit pre- and post-authorization)**. This solution increases maintainability...

```suggestion
Because the first solution (Implicit pre- and post-authorization) increases maintainability and is faster in most cases, always annotate your REST endpoints with the corresponding ``EnforceRoleInResource`` annotation. Always use the annotation for the minimum role that has access.
```

22. REST endpoint best practices for authorization
==================================================

To reject unauthorized requests as early as possible, Artemis employs two solutions:

#. Implicit pre- and post-authorization annotations:
    #. ``EnforceRoleInResource`` (e.g. ``EnforceAtLeastInstructorInCourse``) annotations are responsible for blocking users with wrong or missing authorization roles without querying the database.

>     #. If necessary, these annotations check for access rights to individual resources within the database via light-weight queries.

Reviewer 2: The first line states that there is no querying and the second one states that there might be light-weight queries. This seams a bit confusing...

+------------------+----------------------------------------+-----------------+----------------+

If, for some reason, you need to deviate from these rules, use ``@ManualConfig``. Use this annotation only if absolutely necessary as it will exclude the endpoint from the automatic authorization tests.

Implicit pre- and post-authorization annotations
------------------------------------------------

The following example makes the call only accessible to ADMIN and INSTRUCTOR users and then checks the access rights to the course in the database:

Do **not** write

.. code-block:: java

    @EnforceAtLeastInstructor
    public ResponseEntity<Void> enableLearningPathsForCourse(@PathVariable long courseId) {
        var course = courseRepository.findById(courseId);
        authCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.INSTRUCTOR, course, null);
        [...]
        return ResponseEntity.ok().build();
    }

Instead, use the following annotation:

.. code-block:: java

    @EnforceAtLeastInstructorInCourse
    public ResponseEntity<Void> enableLearningPathsForCourse(@PathVariable long courseId) {
        [...]
        return ResponseEntity.ok().build();
    }

> Utilizing the ``EnforceRoleInResource`` annotation reduces the amount of code and makes the code more readable and maintainable. Also it is faster than the explicit authorization check since we don't have to load the user and course into JAVA as the check is performed directly in a database query.

Reviewer: ```suggestion
Utilizing the ``EnforceRoleInResource`` annotation reduces the amount of code and makes the code more readable and maintainable. Also it is faster than the explicit authorization check since we don't have to load the user and course into Java as the check is performed directly in a database query.
```

Explicit authorization checks
-----------------------------

CAUTION: Be aware that this solution should be used only in those two cases:
    #. when you need to load user **AND** the resource anyway,
    #. when no matching ``EnforceRoleInResource`` annotation exists.

> Always annotate your REST endpoints with the annotation for the minimum role that has access.

Reviewer 2: This was already stated above

> Always annotate your REST endpoints with the corresponding annotation. Always use the annotation for the minimum role that has access.

Reviewer: ```suggestion
Always annotate your REST endpoints with the annotation for the minimum role that has access.
```
